message_ix_models/model/water/data/water_for_ppl.py

```diff
@@ -171,6 +177,91 @@ def hist_cap(x: pd.Series, context: "Context", hold_cost: pd.DataFrame) -> list:
     ]
 
 
+def relax_growth_constraint(ref_hist, scen, cooling_df, g_lo, constraint_type):
+    """
+    Function to check if the parent technologies are shut down and require
+    relaxing the growth constraint.
+
+    Parameters:
+    ref_hist (pd.DataFrame): Historical data.
+    scen (Scenario): Scenario object to retrieve parameters.
+    cooling_df (pd.DataFrame): DataFrame containing cooling technologies and
+    their parent technologies.
+    g_lo (pd.DataFrame): DataFrame containing growth constraints.
+    constraint_type (str): Type of constraint to check ("activity" or "capacity").
+
+    Returns:
+    pd.DataFrame: Updated g_lo DataFrame with relaxed growth constraints.
+    """
+    if constraint_type == "activity":
+        year_type = "year_act"
+        bound_param = "bound_activity_up"
+    elif constraint_type == "new_capacity":
+        year_type = "year_vtg"
+        bound_param = "bound_new_capacity_up"
+    else:
+        raise ValueError(
+            "Invalid constraint_type. Must be 'activity' or 'new_capacity'."
+        )
+
+    # keep rows with max year_type
+    max_year_hist = (
+        ref_hist.loc[ref_hist.groupby(["node_loc", "technology"])[year_type].idxmax()]
+        .drop(columns="unit")
+        .rename(columns={year_type: "hist_year", "value": "hist_value"})
+    )
+
+    # Step 2: Check for bound_activity_up or bound_new_capacity_up conditions
+    bound_up_pare = scen.par(bound_param, {"technology": cooling_df["parent_tech"]})
+    # Get a set with unique year_type values and order them
+    years = np.sort(bound_up_pare[year_type].unique())
+
+    # In max_year_hist add the next year from years matching the hist_year columns
+    max_year_hist["next_year"] = max_year_hist["hist_year"].apply(
+        lambda x: years[years > x][0]
+    )
+
+    # Merge the max_year_hist with bound_up_pare
+    bound_up = pd.merge(bound_up_pare, max_year_hist, how="left")
+    # Look at years just after the historical year
+    bound_up1 = bound_up[bound_up[year_type] == bound_up["next_year"]]
+    # Categories that might break the growth constraints
+    bound_up1 = bound_up1[bound_up1["value"] < 0.9 * bound_up1["hist_value"]]
+    # not look ad sudden contraints after sthe starting year
+    bound_up = bound_up.sort_values(by=["node_loc", "technology", year_type])
+    # Check if value for a year is greater than the value of the next year
+    bound_up["next_value"] = bound_up.groupby(["node_loc", "technology"])[
+        "value"
+    ].shift(-1)
```

Comment:
Just a word of caution: "off by 1"-errors are among the most prominent errors you can make while coding. So I hope you have some kind of test that ensures that shifting by this "magic number" really does exactly what you want it to do.

Reply:
here the shift is not on numbers, but on positions (rows) basically i am taking the 'previous' year values. And since the time steps are not constant I thought it is better then doing -5 or -10. 
However, if there are better suggestions on how to do it, welcome